neurodsp/aperiodic/dfa.py

```diff
@@ -0,0 +1,145 @@
+"""Fluctuation analyses to measure fractal properties of time series."""
+
+import numpy as np
+import scipy as sp
+
+###################################################################################################
+###################################################################################################
+
+def compute_fluctuations(sig, fs, n_scales=10, min_scale=0.01, max_scale=1.0, deg=1, method='dfa'):
+    """Compute a fluctuation analysis on a signal.
+
+    Parameters
+    ----------
+    sig : 1d array
+        Time series.
+    fs : float, Hz
+        Sampling rate, in Hz.
+    n_scales : int, optional, default=10
+        Number of scales to estimate fluctuations over.
+    min_scale : float, optional, default=0.01
+        Shortest scale to compute over, in seconds.
+    max_scale : float, optional, default=1.0
+        Longest scale to compute over, in seconds.
+    deg : int, optional, default=1
+        Polynomial degree for detrending. Only used for DFA.
+
+        - 1 for regular DFA
+        - 2 or higher for generalized DFA
+    method : {'dfa', 'rs'}
+        Method to use to compute fluctuations:
+
+        - 'dfa' : detrended fluctuation
+        - 'rs' : rescaled range
+
+    Returns
+    -------
+    t_scales : 1d array
+        Time-scales over which fluctuation measures were computed.
+    fluctuations : 1d array
+        Average fluctuation at each scale.
+    exp : float
+        Slope of line in loglog when plotting t_scales against fluctuations.
+        This is the alpha value for DFA, or the Hurst exponent for rescaled range.
+
+    Notes
+    -----
+    These analysis compute fractal properties by analyzing fluctuations across windows.
+
+    Overall, these approaches involve dividing the time-series into non-overlapping
+    windows at log-spaced scales and computing a fluctuation measure across windows.
+    The relationship of this fluctuation measure across window sizes provides
+    information on the fractal properties of a signal.
+
+    Measures available are:
+    - DFA: detrended fluctuation analysis
+        - computes ordinary least squares fits across signal windows
+    - RS: rescaled range
+        - computes the range of signal windows, divided by the standard deviation
+
+    Empirically, DFA seems to work for a larger range of exponent values.
+    """
+
+    # Get log10 equi-spaced scales and translate that into window lengths
+    t_scales = np.logspace(np.log10(min_scale), np.log10(max_scale), n_scales)
+    win_lens = np.round(t_scales * fs).astype('int')
+
+    # Calculate culmulative sum of the signal
```

Comment:
I suggest that we move this line into each of the sub-functions, and pass in the original signal to them. It's a copied line, but it feels like it's really part of the computation (and I think it cleans up re-computing the signal in rescaled-range). 

Partly, this makes the the sub-functions more stand-alone, as you pass in an actual time series, and wouldn't need to pre-compute cumulative sum if trying to use directly.

Reply:
Okay sounds good - moved the cumulative sum calculation into the sub functions.